c7n/resources/asg.py

```diff
@@ -1690,6 +1691,59 @@ def __call__(self, config):
         return config['LaunchConfigurationName'] not in self.used
 
 
+@LaunchConfig.filter_registry.register('user-data')
+class UserDataFilter(ValueFilter, LaunchConfigFilterBase):
+
+    schema = type_schema('user-data', rinherit=ValueFilter.schema)
+    batch_size = 50
+    annotation = 'c7n:user-data'
+
+    def validate(self):
+        return self
+
+    def get_permissions(self):
+        return self.manager.get_resource_manager('asg').get_permissions()
+
+    def process(self, launch_configs, event=None):
+        self.data['key'] = '"c7n:user-data"'
+        client = utils.local_session(self.manager.session_factory).client('autoscaling')
+        results = []
+
+        with self.executor_factory(max_workers=3) as w:
+            futures = {}
+            for instance_set in utils.chunks(launch_configs, self.batch_size):
+                futures[w.submit(
+                    self.process_instance_set,
+                    client, instance_set)] = instance_set
+
+            for f in as_completed(futures):
+                if f.exception():
+                    self.log.error(
+                        "Error processing userdata on instance set %s", f.exception())
+                results.extend(f.result())
+        return results
+
+    def process_instance_set(self, client, launch_configs):
+        results = []
+        for lc in launch_configs:
```

Comment:
this looks odd, you've got a launch config, and your fetching the launch config again?

Reply:
Thanks for pointing out, but isn't this simply processing a subset/batch of launch configs I already fetched? The original PR for ec2 userdata did this as well so I kept it.

Reply:
two lines below, its invoking the api per launch config to refetch the same launch config,  which doesn't make any sense.